File: 94_Binary_Tree_Inorder_Traversal.py
'''
Morris Traversal
可以利用常量空间O(1)实现而不借用任何堆栈结构,时间复杂度为O(2N)，依然为线性复杂度, 利用N+1个指向NULL的空指针进行中序遍历。
将一棵普通的二叉树中任一节点的空右孩子指针 指向 中序遍历的后继节点， 空左孩子指针指向中序遍历的前继节点。
实际上就是利用那些空指针，使其指向节点中序遍历的后继节点，前继节点。

Morris遍历方法用了线索二叉树，这个方法不需要为每个节点额外分配指针指向其前驱和后继结点，而是利用叶子节点中的右空指针指向中序遍历下的后继节点就可以了。
算法具体分情况如下：
1. 如果当前结点的左孩子为空，则输出当前结点并将其当前节点赋值为右孩子。
2. 如果当前节点的左孩子不为空，则寻找当前节点在中序遍历下的前驱节点（也就是当前结点左子树的最右孩子）。接下来分两种情况：
 a) 如果前驱节点的右孩子为空，将它的右孩子设置为当前节点（做线索使得稍后可以重新返回父结点）。然后将当前节点更新为当前节点的左孩子。
 b) 如果前驱节点的右孩子为当前节点，表明左子树已经访问完，可以访问当前节点。将它的右孩子重新设为空（恢复树的结构）。输出当前节点。当前节点更新为当前节点的右孩子。 
'''


class Solution:
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = list()
        cur = root
        pre = None
        while cur:
            if not cur.left:
                res.append(cur.val)
                cur = cur.right
            else:
                pre = cur.left
                while pre.right and pre.right != cur:
                    pre = pre.right
                if not pre.right:
                    pre.right = cur
                    cur = cur.left
                else:
                    pre.right = None
                    res.append(cur.val)
                    cur = cur.right
                    
        return res

File: 94.binary-tree-inorder-traversal.go
/*
 * @lc app=leetcode id=94 lang=golang
 *
 * [94] Binary Tree Inorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
	// Approach 1:
	result := []int{}

	var inorder func(*TreeNode)

	inorder = func(root *TreeNode) {
		if root == nil {
			return
		}
		inorder(root.Left)
		result = append(result, root.Val)
		inorder(root.Right)
	}
	inorder(root)

	return result

	// Approach 2: Iterating method using Stack
	// var (
	// 	stack  []*TreeNode
	// 	result []int
	// 	cur    = root
	// )

	// for len(stack) > 0 || cur != nil {
	// 	for cur != nil {
	// 		stack = append(stack, cur)
	// 		cur = cur.Left
	// 	}
	// 	cur = stack[len(stack)-1]
	// 	stack = stack[:len(stack)-1]
	// 	result = append(result, cur.Val)
	// 	cur = cur.Right
	// }
	// return result

}

// @lc code=end

